server/app/views/admin/programs/ProgramBlockPredicateConfigureView.java

```diff
@@ -0,0 +1,621 @@
+                          Optional.of(andNode))));
+    } else {
+      valueRowContainer.with(
+          renderValueRow(questionDefinitions, 1, /* maybeAndNode= */ Optional.empty()));
+    }
+
+    formTag.with(valueRowContainer);
+    formTag.with(div(makeSvgTextButton("Add values", Icons.ADD).withId("predicate-add-value-row")));
+    formTag.with(submitButton("Save condition").withClasses("my-4"));
+
+    return formTag;
+  }
+
+  private static DivTag renderConfiguratorEligibilityHeader() {
+    return div(
+            text("Applicant is eligible when:"),
+            input()
+                .withType("hidden")
+                .withName("predicateAction")
+                .withValue(PredicateAction.ELIGIBLE_BLOCK.name()))
+        .withClasses("font-bold");
+  }
+
+  private static DivTag renderConfiguratorVisibilityHeader(
+      Optional<PredicateDefinition> maybeExistingPredicate) {
+    DivTag container =
+        div("Screen is ").withClasses("font-bold", ReferenceClasses.PREDICATE_ACTION);
+    SelectTag selectTag = select().withName("predicateAction");
+
+    ImmutableList.of(PredicateAction.HIDE_BLOCK, PredicateAction.SHOW_BLOCK)
+        .forEach(
+            action ->
+                selectTag.with(
+                    option(action.toDisplayString())
+                        .withClasses(
+                            ReferenceClasses.MULTI_OPTION_QUESTION_OPTION,
+                            ReferenceClasses.MULTI_OPTION_VALUE)
+                        .withCondSelected(
+                            maybeExistingPredicate
+                                .map(PredicateDefinition::action)
+                                .map(existingAction -> existingAction.equals(action))
+                                .orElse(false))
+                        .withValue(action.name())));
+
+    container.with(selectTag);
+
+    return container;
+  }
+
+  private ImmutableList<PredicateExpressionNode> getExistingAndNodes(
+      PredicateDefinition existingPredicate) {
+    switch (existingPredicate.computePredicateFormat()) {
+      case SINGLE_QUESTION:
+        {
+          return ImmutableList.of(
+              PredicateExpressionNode.create(
+                  AndNode.create(ImmutableList.of(existingPredicate.rootNode()))));
+        }
+
+      case OR_OF_SINGLE_LAYER_ANDS:
+        {
+          return existingPredicate.rootNode().getOrNode().children();
+        }
+
+      default:
+        {
+          throw new RuntimeException(
+              String.format(
+                  "Unrecognized predicate format: %s", existingPredicate.computePredicateFormat()));
+        }
+    }
+  }
+
+  private DivTag renderQuestionHeaders(
+      ImmutableList<QuestionDefinition> questionDefinitions,
+      Optional<PredicateDefinition> maybeExistingPredicate) {
+    DivTag container = div().withClasses("flex", "py-4");
+    int columnNumber = 1;
+
+    if (maybeExistingPredicate.isPresent()) {
```

Comment:
You effectively have two methods here, renderExisting and renderNew.

Suggestion to break it into two methods, or perhaps make the isEmpty() first and return early, which unindents most of this method which aids readability.

Reply:
an early return creates more code duplication between them because of the empty div needed to keep all the columns aligned
```java
    return container.with(div().withClasses("w-28"));
```

Reply:
matter of preference I suppose.  duplicating the two lines doesn't seem like bad duplication to me for the readability, though doesn't make the duplication good either, but also you could pass it in to the two methods.  lgtm eitheway